### PR TITLE
update type annotations

### PR DESCRIPTION
**Pull request by danieleades**
- replaced a few type comments with type annotations

wanted to check this was of interest before i got too deep into this

### CODE REVIEW

**Comment by frostming** [comment]
Glad to see this piece of work is taken and started. I will be even better if a step can be added to `Lint` CI that invokes `mypy` in order to see the effect of the improvement.

**Comment by danieleades** [comment]
> Glad to see this piece of work is taken and started. It will be even better if a step can be added to `Lint` CI that invokes `mypy` in order to see the effect of the improvement.

Cool, I'll keep going.
I'd like to add mypy linting too, that's my eventual goal. but there's a way to go before that!

**Comment by danieleades** [comment]
that's the rest of the type comments replaced with annotations.

mypy is currently returning >500 errors. Will probably start chipping away at those in separate pull requests, and keep this one focused on replacing the type comments only.

File: pdm/models/specifiers.py [hunk 1]
```diff
@@ -82,17 +82,17 @@ def __init__(self, version_str: str = "", analyze: bool = True) -> None:
...
+        self._excludes: List[Tuple[Union[int, str], ...]] = []
         if version_str and analyze:
             self._analyze_specifiers()
 
     def _analyze_specifiers(self) -> None:
         # XXX: Prerelease or postrelease specifiers will fail here, but I guess we can
         # just ignore them for now.
         lower_bound, upper_bound = self.MIN_VERSION, self.MAX_VERSION
-        excludes = set()  # type: Set[Tuple[Union[int, str], ...]]
+        excludes: Set[Tuple[Union[int, str], ...]] = set()
```
Comment (danieleades) [inline pdm/models/specifiers.py:95]: might be nice to tidy this up with a type alias? what exactly is represented by a `Tuple[Union[int, str], ...]` here?

Comment (frostming) [reply pdm/models/specifiers.py:95]: Each item is a version tuple, which accepts the following forms:

- `(3, 8, 0)`
- `(3, 8, '*')`

**Comment by frostming** [comment]
> I'd like to add mypy linting too, that's my eventual goal. but there's a way to go before that!

No hurry, just a reminder though.

> Will probably start chipping away at those in separate pull requests, and keep this one focused on replacing the type comments only.

That is okay, will start reviewing later

File: pdm/models/specifiers.py (hunk 1 above)
Comment (danieleades) [reply pdm/models/specifiers.py:95]: ah I see, so something along the lines of 

```python
VersionIncrement = Union[int, Literal['*']]
SemVer = Tuple[VersionIncrement, VersionIncrement, VersionIncrement]

...

excludes: Set[SemVer]
```

I'll have think about a tidy way to capture this.

almost worth a class.

Comment (danieleades) [reply pdm/models/specifiers.py:95]: i feel like `pdm/models/specifiers.py` is crying out for some custom classes to encapsulate the version specifications. The code here looks challenging to debug/maintain

like this - https://docs.rs/semver/0.11.0/semver/struct.Version.html
and this - https://docs.rs/semver/0.11.0/semver/struct.VersionReq.html

I'm not planning to touch that in this pull request. will resolve this conversation for now

Comment (frostming) [reply pdm/models/specifiers.py:95]: Looks good. But be noted Literal is not supported in Python 3.7

**Review by frostming — CHANGES_REQUESTED** [review]
LGTM! Would you mind adding a news entry `298.misc.md` describing what you did in this PR?

**Comment by danieleades** [comment]
> LGTM! Would you mind adding a news entry `news/298.misc.md` describing what you did in this PR?

sure. Is that what you were after?

(i've not used `towncrier` before)

**Review by frostming — APPROVED** [review]

**Comment by frostming** [comment]
No more to do, that is it

**Pull request merged**